tools/functional-tester/etcd-tester/failure.go

```diff
@@ -173,6 +201,105 @@ func (f *failureKillOneForLongTime) Recover(c *cluster, round int) error {
 	return c.WaitHealth()
 }
 
+// failureKillLeaderForLongTime kills the leader for long time, and restart
+// after a snapshot is required.
+type failureKillLeaderForLongTime struct {
+	description
+	idx int
+}
+
+func newFailureKillLeaderForLongTime() *failureKillLeaderForLongTime {
+	return &failureKillLeaderForLongTime{
+		description: "kill the leader for long time and expect it to recover from incoming snapshot",
+	}
+}
+
+func (f *failureKillLeaderForLongTime) Inject(c *cluster, round int) error {
+	idx, leaderEndpoint := c.GetLeader()
+	f.idx = idx
+	if leaderEndpoint == "" {
+		log.Println("etcd-tester: no leader found; just stopping", c.ClientURLs[idx])
+	}
+	if err := c.Agents[idx].Stop(); err != nil {
+		return err
+	}
+	if c.Size >= 3 {
+		start, _ := c.Report()
+		var end int
+		retry := snapshotCount / 1000 * 3
+		for j := 0; j < retry; j++ {
+			end, _ = c.Report()
+			if end-start > snapshotCount {
+				return nil
+			}
+			time.Sleep(time.Second)
+		}
+		return fmt.Errorf("cluster too slow: only commit %d requests in %ds", end-start, retry)
+	}
+	return nil
+}
+
+func (f *failureKillLeaderForLongTime) Recover(c *cluster, round int) error {
+	if _, err := c.Agents[f.idx].Restart(); err != nil {
+		return err
+	}
+	return c.WaitHealth()
+}
+
+// failureKillLeaderForLongTimeAndAnother kills the leader for a long time
+// and kills one of the rest.
+type failureKillLeaderForLongTimeAndAnother struct {
```

Comment:
this test behaviors differently with what happened in case #3. In case 3, the second error happens before the cluster receives enough commits. In this test, it injects the error after the cluster receives enough commits. 

In another, I would like to see a clear goal for adding this kind of complicated failure injections.